src/Bicep.Core/Emit/SourceMapHelper.cs

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+using Bicep.Core.Extensions;
+using Bicep.Core.FileSystem;
+using Bicep.Core.Parsing;
+using Bicep.Core.Syntax;
+using Bicep.Core.Text;
+using Bicep.Core.Workspaces;
+using Microsoft.WindowsAzure.ResourceStack.Common.Extensions;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Bicep.Core.Emit
+{
+
+    public record SourceMap(
+        string Entrypoint,
+        ImmutableArray<SourceMapFileEntry> Entries);
+
+    public record SourceMapFileEntry(
+        string FilePath,
+        ImmutableArray<SourceMapEntry> SourceMap);
+
+    public record SourceMapEntry(
+        int SourceLine,
+        int TargetLine);
+
+    public record RawSourceMap(
+        IList<RawSourceMapFileEntry> Entries);
+
+    public record RawSourceMapFileEntry(
+        string FilePath,
+        IList<SourceMapRawEntry> SourceMap);
+
+    public record SourceMapRawEntry(
+        TextSpan SourcePosition,
+        IList<TextSpan> TargetPositions);
+
+    public static class SourceMapHelper
+    {
+        private static readonly Regex JsonWhitespaceStrippingRegex = new(@"(""(?:[^""\\]|\\.)*"")|\s+", RegexOptions.Compiled);
+
+        public static void AddMapping(
+            RawSourceMap rawSourceMap,
+            BicepFile bicepFile,
+            SyntaxBase bicepSyntax,
+            PositionTrackingJsonTextWriter jsonWriter,
+            int jsonStartPos)
+        {
+            if (rawSourceMap is null)
+            {
+                throw new ArgumentNullException(nameof(rawSourceMap));
+            }
+
+            var bicepFilePath = bicepFile.FileUri.AbsolutePath;
+            TextSpan bicepLocation = bicepSyntax.Span;
+
+            // account for leading nodes (decorators)
+            if (bicepSyntax is StatementSyntax syntax)
+            {
+                var lastLeadingNode = syntax.LeadingNodes.LastOrDefault();
+                if (lastLeadingNode is not null)
+                {
+                    bicepLocation = new TextSpan(
+                        lastLeadingNode.Span.Position + lastLeadingNode.Span.Length,
+                        syntax.Span.Length - lastLeadingNode.Span.Length);
+                }
+            }
```

Comment:
The output in Bicep.Core.Samples is great for having a high level overview of whether things are working as expected, but it'll be difficult for people without context to understand whether a particular change is going to have significant impact - meaning it's possible we'll have small regressions.

In the long run it might be worth thinking about adding some unit tests for this functionality as well - logic such as this is going to be a bit fragile if e.g. syntax changes are made by someone without the context on source mapping. The naming of the tests would also establish intent for anyone unfamiliar - e.g. "Sourcemap_output_should_account_for_decorators_in_statement_syntax".

Reply:
Per our offline discussion this should not be blocking for this PR. I will create a follow up PR with more unit tests.